### PR TITLE
chore: pin actions to digests and introduce step security hardened runners

### PR DESCRIPTION
**Pull request by jedsalazar**
Pin GitHub Actions to a digest to [prevent supply chain attacks](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions). Adds Step Security Github Actions Harden Runner agent, which monitors for least privilege GITHUB_TOKEN permissions and egress connections. More info [here](https://github.com/step-security/harden-runner)

### CODE REVIEW

**Review by azenla — APPROVED** [review]
LGTM

File: .github/workflows/release-binaries.yml [hunk 1]
```diff
@@ -72,16 +76,20 @@ jobs:
...
         submodules: recursive
-    - uses: dtolnay/rust-toolchain@stable
+    - uses: dtolnay/rust-toolchain@d388a4836fcdbde0e50e395dc79a2670ccdef13f # stable
       if: ${{ matrix.platform.os != 'darwin' }}
     - uses: dtolnay/rust-toolchain@stable
       with:
         targets: "${{ matrix.platform.arch }}-apple-darwin"
       if: ${{ matrix.platform.os == 'darwin' }}
-    - uses: homebrew/actions/setup-homebrew@master
+    - uses: homebrew/actions/setup-homebrew@4b34604e75af8f8b23b454f0b5ffb7c5d8ce0056 # master
```
Comment (khionu) [inline .github/workflows/release-binaries.yml:92]: One more version comment

**Review by khionu — CHANGES_REQUESTED** [review]
Missed a handful of version comments, that's it

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -11,10 +11,14 @@ jobs:
...
-    - uses: actions/checkout@v4
+    - name: Harden Runner
+      uses: step-security/harden-runner@a4aa98b93cab29d9b1101a6143fb8bce00e2eac4 # v2.7.1
+      with:
+        egress-policy: audit
+    - uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
       with:
         submodules: recursive
-    - uses: dtolnay/rust-toolchain@stable
+    - uses: dtolnay/rust-toolchain@d388a4836fcdbde0e50e395dc79a2670ccdef13f # stable
```
[the remark this answers is not in the document]
Comment (jedsalazar) [reply .github/workflows/check.yml:21]: I can't find the latest stable explicit version. They [don't publish releases](https://github.com/dtolnay/rust-toolchain)

File: .github/workflows/release-binaries.yml (hunk 1 above)
Comment (jedsalazar) [reply .github/workflows/release-binaries.yml:92]: Apparently they only publish master. The idea here would be to pin to master and then let Dependabot upgrade and give us the code review diffs and cadence we prefer. Let me know what you think

**Review by kaniini — APPROVED** [review]

File: .github/workflows/release-binaries.yml (hunk 1 above)
Comment (khionu) [reply .github/workflows/release-binaries.yml:92]: In lieu of actual versions, that seems like the best choice

File: .github/workflows/check.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (khionu) [reply .github/workflows/check.yml:21]: Apparently they use the version field to allow you to pick the Rust version

> The selection of Rust toolchain is made based on the particular @rev of this Action being requested. For example "dtolnay/rust-toolchain@nightly" pulls in the nightly Rust toolchain, while "dtolnay/rust-toolchain@1.42.0" pulls in 1.42.0.

This makes me a bit uneasy as the action itself can't be versioned properly with this approach, except by, as you did, pinning the commit. I'm curious how the action will behave given it tries to use that value itself... but I think it'll be the best use of our time if we just let it play out. We'll know what to fix if it breaks, and it won't block much for long.

**Review by khionu — APPROVED** [review]

**Pull request merged**